src/Collection/Form/CollectionTypeExtension.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace Symfony\UX\Collection\Form;
+
+use Symfony\Component\Form\AbstractTypeExtension;
+use Symfony\Component\Form\Extension\Core\Type\ButtonType;
+use Symfony\Component\Form\Extension\Core\Type\CollectionType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\Options;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class CollectionTypeExtension extends AbstractTypeExtension
+{
+    public static function getExtendedTypes(): iterable
+    {
+        return [CollectionType::class];
+    }
+
+
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        /** @var FormInterface|null $prototype */
+        $prototype = $builder->getAttribute('prototype');
+
+        if (!$prototype) {
+            return;
+        }
+
+        // TODO add button only if `delete_type` is defined and set `delete_type` default to null?
+        if ($options['allow_delete']) {
+            // add delete button to prototype
+            // TODO add toolbar here to allow extension add other buttons
+            $prototype->add('deleteButton', $options['delete_type'], $options['delete_options']);
+        }
+    }
+
+    public function buildView(FormView $view, FormInterface $form, array $options): void
+    {
+        /** @var FormInterface|null $prototype */
+        $prototype = $form->getConfig()->getAttribute('prototype');
+
+        if (!$prototype) {
+            return;
+        }
+
+        if ($options['allow_delete']) {
+            // add delete button to rendered elements from the Collection ResizeListener
+            foreach ($form as $child) {
+                $child->add('deleteButton', $options['delete_type'], $options['delete_options']);
```

Comment:
I get an `AlreadySubmittedException` when the form is submitted and shown again (e. g. due to validation errors).

> You cannot add children to a submitted form.

I think the solution is to add the buttons in an `FormEvents::PRE_SET_DATA` event listener in `buildForm()`:


```php
$builder->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event) use ($options) {
    foreach ($event->getForm() as $name => $child) {
        $child->add('deleteButton', $options['delete_type'], $options['delete_options']);
    }

    $event->getForm()->add('addButton', $options['add_type'], $options['add_options']);
});
```

Reply:
The add button needs to be added to the form again on submit because it is removed in the `ResizeFormListener`.

```php
$builder->addEventListener(FormEvents::SUBMIT, function (FormEvent $event) use ($options) {
    $form = $event->getForm();

    if ($form->has('addButton')) {
        return;
    }

    $form->add('addButton', $options['add_type'], $options['add_options']);
});
```